File: sdk/python/dogdata/resources/forensic.py
from __future__ import annotations

from typing import Any, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from dogdata.client import DogData


class ForensicResource:
    def __init__(self, client: DogData):
        self._client = client

    def profiles(self, page: int = 1, limit: int = 100) -> dict[str, Any]:
        return self._client.request("/api/forensic/profiles", {
            "page": page,
            "limit": limit,
        })

    def profile(self, address: str) -> dict[str, Any]:
        return self._client.request("/api/forensic/profiles", {
            "address": address,
        })

    def summary(self) -> dict[str, Any]:
        return self._client.request("/api/forensic/summary")

    def by_category(self, category: str, page: int = 1, limit: int = 100) -> dict[str, Any]:
        return self._client.request("/api/forensic/profiles", {
            "category": category,
            "page": page,
            "limit": limit,
        })

    def diamond_hands(self, page: int = 1, limit: int = 100) -> dict[str, Any]:
        return self.by_category("diamond_hands", page, limit)

    def paper_hands(self, page: int = 1, limit: int = 100) -> dict[str, Any]:
        return self.by_category("paper_hands", page, limit)

File: sdk/typescript/src/resources/forensic.ts
import type { DogData } from '../client.js';
import type { PaginationParams, PaginatedResponse, ForensicProfile, ForensicSummary } from '../types.js';

export class ForensicResource {
  constructor(private client: DogData) {}

  async profiles(params?: PaginationParams): Promise<PaginatedResponse<ForensicProfile>> {
    return this.client.request<PaginatedResponse<ForensicProfile>>('/api/forensic/profiles', {
      page: params?.page,
      limit: params?.limit,
    });
  }

  async profile(address: string): Promise<ForensicProfile> {
    return this.client.request<ForensicProfile>('/api/forensic/profiles', {
      address,
    });
  }

  async summary(): Promise<ForensicSummary> {
    return this.client.request<ForensicSummary>('/api/forensic/summary');
  }

  async byCategory(category: string, params?: PaginationParams): Promise<PaginatedResponse<ForensicProfile>> {
    return this.client.request<PaginatedResponse<ForensicProfile>>('/api/forensic/profiles', {
      category,
      page: params?.page,
      limit: params?.limit,
    });
  }

  async diamondHands(params?: PaginationParams): Promise<PaginatedResponse<ForensicProfile>> {
    return this.byCategory('diamond_hands', params);
  }

  async paperHands(params?: PaginationParams): Promise<PaginatedResponse<ForensicProfile>> {
    return this.byCategory('paper_hands', params);
  }
}
